provider/resource_keycloak_default_groups_test.go

```diff
@@ -0,0 +1,231 @@
+					found = true
+				}
+			}
+
+			if !found {
+				return fmt.Errorf("unable to find group %s", groupName)
+			}
+		}
+
+		return nil
+	}
+}
+
+func testAccCheckGroupsArentDefault(resourceName string, groupNames []string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		stateGroups, err := testAccGetGroupsFromDefaultGroup(resourceName, s)
+		if err != nil {
+			return err
+		}
+
+		for _, stateGroup := range stateGroups {
+			for _, groupName := range groupNames {
+				if groupName == stateGroup.Name {
+					return fmt.Errorf("didnt expect to find group %s", groupName)
+				}
+			}
+		}
+
+		return nil
+	}
+}
+
+func testAccNoDefaultGroups(resourceName string, groupNames []string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		stateGroups, err := testAccGetGroupsFromDefaultGroup(resourceName, s)
+		if err != nil {
+			return err
+		}
+
+		if len(stateGroups) != 0 {
+			return fmt.Errorf("expected 0 stateGroups, got %d", len(stateGroups))
+		}
+
+		return nil
+	}
+}
+
+func testAccGetGroupsFromDefaultGroup(resourceName string, s *terraform.State) ([]keycloak.Group, error) {
+	keycloakClient := testAccProvider.Meta().(*keycloak.KeycloakClient)
+
+	rs, ok := s.RootModule().Resources[resourceName]
+	if !ok {
+		return nil, fmt.Errorf("resource not found: %s", resourceName)
+	}
+
+	realmId := rs.Primary.Attributes["realm_id"]
+
+	return keycloakClient.GetDefaultGroups(realmId)
+}
+
+func testKeycloakDefaultGroups_basic(realmName, groupName string) string {
+	return fmt.Sprintf(`
+resource "keycloak_realm" "realm" {
+	realm = "%s"
+}
+
+resource "keycloak_group" "group" {
+	name     = "%s"
+	realm_id = "${keycloak_realm.realm.id}"
+}
+
+resource "keycloak_default_groups" "group_default" {
+	realm_id = "${keycloak_realm.realm.id}"
+	group_ids = ["${keycloak_group.group.id}"]
+}
+	`, realmName, groupName)
+}
+
+func testKeycloakDefaultGroups_noDefaultGroups(realmName, groupName string) string {
+	return fmt.Sprintf(`
+resource "keycloak_realm" "realm" {
+	realm = "%s"
+}
+
+resource "keycloak_group" "group" {
+	name     = "%s"
+	realm_id = "${keycloak_realm.realm.id}"
+}`, realmName, groupName)
+}
+
+// this tf config provides a good way to test groups that exist within keycloak but aren't defaults
+func testKeycloakDefaultGroups_multipleGroups(realm string, groups []string, defaultGroups []string) string {
+	out := fmt.Sprintf(`
+resource "keycloak_realm" "realm" {
+	realm = "%s"
+}`, realm)
+
+	for _, group := range groups {
+		out += fmt.Sprintf(`
+resource "keycloak_group" "%s" {
+	name     = "%s"
+	realm_id = "${keycloak_realm.realm.id}"
+}`, group, group)
+	}
+
+	var defaultGroupResources []string
+	for _, defaultGroup := range defaultGroups {
+		defaultGroupResources = append(defaultGroupResources, fmt.Sprintf(`"${keycloak_group.%s.id}"`, defaultGroup))
+	}
+
+	out += fmt.Sprintf(`
+resource "keycloak_default_groups" "group_default" {
+	realm_id = "${keycloak_realm.realm.id}"
+	group_ids = [%s]
```

Comment:
I have a helper function in the test package called `arrayOfStringsForTerraformResource` that does this for you, could you refactor this to use that function?

Reply:
Perfect! Can do.